src/main/java/org/gradle/java/JigsawPlugin.java

```diff
@@ -0,0 +1,231 @@
+        final String moduleDir =
+                compileJava.getDestinationDir().toString() + System.getProperty("path.separator") + module.getModuleName();
+        compileJava.getInputs().property("moduleName", module.getModuleName());
+        compileJava.doFirst(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                List<String> args = new ArrayList<>();
+                args.add("--module-path");
+                args.add(compileJava.getClasspath().getAsPath());
+                args.add("-d");
+                args.add(moduleDir);
+                compileJava.getOptions().setCompilerArgs(args);
+                compileJava.setClasspath(project.files());
+            }
+        });
+        compileJava.doLast(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                project.copy(new Action<CopySpec>() {
+                    @Override
+                    public void execute(CopySpec copySpec) {
+                        copySpec.from(moduleDir);
+                        copySpec.into(compileJava.getDestinationDir());
+                    }
+                });
+            }
+        });
+    }
+
+    private void configureCompileTestJavaTask(final Project project) {
+        final JavaCompile compileTestJava = (JavaCompile) project.getTasks()
+                .findByName(JavaPlugin.COMPILE_TEST_JAVA_TASK_NAME);
+        final SourceSet test = ((SourceSetContainer) project.getProperties().get("sourceSets")).getByName("test");
+        final Architecture module = (Architecture) project.getExtensions().getByName(EXTENSION_NAME);
+        compileTestJava.getInputs().property("moduleName", module.getModuleName());
+        compileTestJava.doFirst(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                List<String> args = new ArrayList<>();
+                args.add("--module-path");
+                args.add(compileTestJava.getClasspath().getAsPath());
+                args.add("--add-modules");
+                args.add("junit");
+                args.add("--add-reads");
+                args.add(module.getModuleName() + "=junit");
+                args.add("--patch-module");
+                args.add(module.getModuleName() + "=" + test.getJava().getSourceDirectories().getAsPath());
+                compileTestJava.getOptions().setCompilerArgs(args);
+                compileTestJava.setClasspath(project.files());
+            }
+        });
+    }
+
+    private void configureJarTask(final Project project) {
+        final Jar jar = (Jar) project.getTasks().findByName(JavaPlugin.JAR_TASK_NAME);
+        final Architecture module = (Architecture) project.getExtensions().getByName(EXTENSION_NAME);
+        jar.getInputs().property("moduleName", module.getModuleName());
+        jar.exclude(module.getModuleName());
```

Comment:
The jar is the wrong place to exclude this. When compiling against a java library we use the classes dir. The classes dir needs to be clean.